src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
+ * Starts a new transaction in the current thread.
+ * If called within an open transaction, starts a nested transaction.
+ *
+ * If successful, transaction stage changes to TX_STAGE_WORK and function
+ * returns zero. Otherwise, stage changes to TX_STAGE_ONABORT and an error
+ * number is returned.
+ */
+int pmemobj_tx_begin(PMEMobjpool *pop, jmp_buf *env, enum tx_lock, ...);
+
+/*
+ * Aborts current transaction
+ *
+ * Must be called during TX_STAGE_WORK. Otherwise, has no effect.
+ *
+ * Always causes transition to TX_STAGE_ONABORT.
+ */
+void pmemobj_tx_abort(int errnum);
+
+/*
+ * Commits current transaction
+ *
+ * If successful and called during TX_STAGE_WORK, transaction stage changes
+ * to TX_STAGE_ONCOMMIT and function returns zero. Otherwise, stage changes
+ * to TX_STAGE_ONABORT and an error number is returned.
+ */
+int pmemobj_tx_commit();
+
+/*
+ * Cleanups current transaction. Must always be called after pmemobj_tx_begin,
+ * even if starting the transaction failed.
+ *
+ * If called during TX_STAGE_NONE, has no effect.
+ *
+ * Always causes transition to TX_STAGE_NONE.
+ */
+void pmemobj_tx_end();
+
+/*
+ * Performs the actions associated with current stage of the transaction,
+ * and makes the transition to the next stage. Current stage must always
+ * be obtained by calling pmemobj_tx_get_stage.
+ *
+ * If successful, function returns zero. Otherwise, an error number is returned.
+ */
+int pmemobj_tx_process();
+
+#define	__TX_BEGIN(pop, locks, ...)					\
+{									\
+	jmp_buf __env;							\
+	pmemobj_tx_begin(pop, &__env, locks, ##__VA_ARGS__);		\
+	setjmp(__env);							\
+	switch ((int)pmemobj_tx_get_stage()) {				\
+		case TX_STAGE_WORK:					\
+
+#define	TX_BEGIN_WRLOCK(pop, ...) __TX_BEGIN(pop, TX_LOCK_RWLOCK, ##__VA_ARGS__)
+
+#define	TX_BEGIN_LOCK(pop, ...) __TX_BEGIN(pop, TX_LOCK_MUTEX, ##__VA_ARGS__)
+
+#define	TX_BEGIN(pop) TX_BEGIN_LOCK(pop)
+
+#define	TX_ONABORT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONABORT:
+
+#define	TX_ONCOMMIT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONCOMMIT:
+
+/* called regardless of the transaction end state */
+#define	TX_FINALLY							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_FINALLY:
+
+#define	TX_END								\
+			pmemobj_tx_process();				\
+			break;						\
+		default:						\
+			pmemobj_tx_process();				\
+			break;						\
+	}								\
+	pmemobj_tx_end();						\
+}
+
+/*
+ * Takes a "snapshot" of the memory block of given size and located at given
+ * offset, and saves it in the undo log.
+ * The application is then free to directly modify the object in that memory
+ * range. In case of failure or abort, all the changes within this range will
+ * be rolled-back automatically.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+int pmemobj_tx_add_range(uint64_t off, size_t size);
```

Comment:
Would it be worth to add a comment describing what is the *off*? I mean, is this the pool offset (poffset) or heap offset (hoffset)? How to calculate such offset? The hint is in `TX_ADD_FIELD` macro, but maybe the same formula should be presented here?